### PR TITLE
Rework WAF Bindings 

### PR DESCRIPTION
**Pull request by eliottness**
Basically writing all the new implementation. Please look at the new `README.md` if you need an overview

```
Purego Copies (branche purego-v2):
BenchmarkOverall
BenchmarkOverall/16
BenchmarkOverall/16-10 	  263692	      4459 ns/op
BenchmarkOverall/64
BenchmarkOverall/64-10 	  255783	      4699 ns/op
BenchmarkOverall/128
BenchmarkOverall/128-10         	  238806	      4984 ns/op
BenchmarkOverall/256
BenchmarkOverall/256-10         	  213303	      5550 ns/op
BenchmarkOverall/512
BenchmarkOverall/512-10         	  180177	      6652 ns/op
BenchmarkOverall/1024
BenchmarkOverall/1024-10        	  128613	      8707 ns/op
BenchmarkOverall/2048
BenchmarkOverall/2048-10        	   97942	     12371 ns/op

CGO (branche main):
BenchmarkOverall
BenchmarkOverall/16
BenchmarkOverall/16-10 	  176250	      6634 ns/op
BenchmarkOverall/64
BenchmarkOverall/64-10 	  184581	      6558 ns/op
BenchmarkOverall/128
BenchmarkOverall/128-10         	  174375	      6719 ns/op
BenchmarkOverall/256
BenchmarkOverall/256-10         	  137481	      8316 ns/op
BenchmarkOverall/512
BenchmarkOverall/512-10         	  145016	      8254 ns/op
BenchmarkOverall/1024
BenchmarkOverall/1024-10        	  140026	      8677 ns/op
BenchmarkOverall/2048
BenchmarkOverall/2048-10        	  126534	      9512 ns/op
```

### CODE REVIEW

File: handle.go
Comment (Julio-Guerra) [inline handle.go:86]: The previous CAS implementation of the handle ref counter is still needed to avoid a race between the context creations and their WAF handle closing. `Load() == 0` isn't enough, and can become true right after it was false in a call to NewContext.

File: handle.go
Comment (Julio-Guerra) [inline handle.go:82]: Sorry, it feels super weird in the end to see `NewContext` as a method 😅
I'd feel more comfortable reverting to a function to get back something more idiomatic.

**Review by Julio-Guerra — CHANGES_REQUESTED** [review]
Quick first pass only mainly on the encoder and allocator.

File: context.go
Comment (Julio-Guerra) [inline context.go:64]: After reviewing the purpose and usage of the allocator, I think this top-level `wafObject` is missing in the list of ref-counted Go values in `encoder.allocator`.

File: context.go
Comment (Julio-Guerra) [inline context.go:29]: why not `encoder` directly?

File: context.go
Comment (Julio-Guerra) [inline context.go:75]: I suggest you don't pass allocator here and simply call KeepAlive from here.

File: handle.go
Comment (Julio-Guerra) [reply handle.go:86]: I added a test covering stressing the ref-counter in https://github.com/DataDog/go-libddwaf/pull/15/commits/8bb0d5b0310c6adebac3f7da214c324f9d5fb7a8 

It works like a charm and finds the issue :-)

File: handle.go
Comment (Hellzy) [inline handle.go:30]:
```suggestion
// - Encode the rule free form object as a ddwaf_object to send them to the WAF
```

File: handle.go
Comment (Hellzy) [inline handle.go:107]:
```suggestion
// CloseContext calls ddwaf_context_destroy and eventually ddwaf_destroy on the handle
```

File: handle.go
Comment (Hellzy) [inline handle.go:115]:
```suggestion
// Close puts the handle in termination state, when all the contexts are closed the handle will be destroyed
```

File: README.md
Comment (eliottness) [reply README.md:109]:
```suggestion
Another requirement because of this to have a FHS filesystem on your machine and, for linux, to provide `libc.so.6`, `libpthread.so.0` and `libm.so.6`, `libdl.so.2` as dynamic libraries.
```

Comment (eliottness) [reply README.md:109]: 👍

File: README.md
Comment (eliottness) [reply README.md:118]:
```suggestion
- Keep in mind that we fully escape the type system. If you send the wrong data it will segfaults in the best cases but not always!
```

File: allocator.go
Comment (eliottness) [reply allocator.go:70]: 😱 This makes my life really easier

File: context.go
Comment (eliottness) [reply context.go:46]: Sure, let's change this
```suggestion
// * actions: A array of string referencing actions from triggered rules (`on_match` rule field)
```

File: encoder.go
Comment (eliottness) [reply encoder.go:197]:
```suggestion
// We use the function allocator.AllocWafMapKey to store the key in the wafObject. But first we need
```

File: allocator.go
Comment (eliottness) [inline allocator.go:51]:
```suggestion
func (allocator *allocator) AllocWafString(obj *wafObject, str string) {
```

File: waf_disabled.go
Comment (eliottness) [reply waf_disabled.go:25]: Yes but we have to at least be able to compile with it since it's part of the API of the package

File: handle.go
Comment (eliottness) [reply handle.go]: done

File: handle.go
Comment (eliottness) [reply handle.go]: It's working :)

File: handle.go
Comment (Julio-Guerra) [inline handle.go:52]: Can we make `InitWaf` private?

File: handle.go
Comment (Julio-Guerra) [inline handle.go:123]: Please make this method private

File: internal/empty-free/empty_free.go
Comment (Julio-Guerra) [inline internal/empty-free/empty_free.go:12]: Let's rename it `NoopFreeFn`:
```suggestion
var NoopFreeFn uintptr = uintptr(unsafe.Pointer(&_empty_free))
```

File: handle.go
Comment (Julio-Guerra) [inline handle.go:168]:
```suggestion
		// Prevent libddwaf from freeing our Go-memory-allocated ddwaf_objects 
		freeFn: emptyfree.EmptyFreeFn,
```

File: handle.go
Comment (eliottness) [reply handle.go:52]: Yes

File: waf.go
Comment (Julio-Guerra) [inline waf.go:41]: As of the current state of `InitWaf`, `Health()` can currently fail if things like `os.Remove` fails, while we must be more permissive and let using the libddwaf as long as it is usable. OTOH, I think all the errors are worth noting via the dd-trace-go logger, which we can't access from here.

For this reason, I suggest we move to a slightly different API where we don't loose the details:
1. Rename `InitWaf()` into `Load()`, and `CloseWaf()` into `Close()`.
2. Change the API of `Load()` into `func Load() (ok bool, err error)` so that we can let the caller can know if libddwaf is usable or not thanks to `ok`:
    1. When `ok` is true, a non-nil error is acceptable to me in order to return such `os.Remove()` errors for instance that we can log in the dd-trace-go side.
    2. When `ok` is false (ie. not usable), the error must state why, and so such `os.Remove()` errors shouldn't take precedence over the actual libddwaf-related errors.

I'm ok doing such changes in a dedicated separate PR.

File: waf_dl.go
Comment (Julio-Guerra) [inline waf_dl.go:57]: I think we should ignore them, or at least keep the error if already not nil.

File: handle.go
Comment (Julio-Guerra) [reply handle.go:52]: Closing this for now as I have new ideas around Health vs InitWaf basically coming in my pending review.

File: handle.go
Comment (eliottness) [reply handle.go]: done

File: allocator.go
Comment (eliottness) [reply allocator.go]: done

File: waf.go
Comment (eliottness) [reply waf.go:41]: Sure let there be a Jira ticket

File: context.go
Comment (Julio-Guerra) [inline context.go:52]: Making it a bit higher level and better looking for godoc:

```suggestion
// Run encodes the given addressesToData values and runs them against the WAF rules within the given
// timeout value. It returns the matches as a JSON string (usually opaquely used) along with the corresponding
// actions in any. In case of an error, matches and actions can still be returned, for instance in the case of a
// timeout error. Errors can be tested against the RunError type.
```

File: ctypes.go
Comment (Julio-Guerra) [inline ctypes.go:163]:
```suggestion
// keepAlive() globals
var (
	alwaysFalse bool
	escapeSink any
)
```

File: ctypes.go
Comment (Julio-Guerra) [inline ctypes.go:169]:
```suggestion
// - It forces the deallocation of the memory to take place later than expected (just like runtime.KeepAlive)
// - It forces the given argument x to be escaped on the heap by saving it into a global value (Go doesn't provide a standard way to do it as of today)
// It is implemented so that the compiler cannot optimize it.
```

File: waf.go
Comment (eliottness) [reply waf.go:38]: done

File: context.go
Comment (Julio-Guerra) [inline context.go:67]:
```suggestion
	// Save the Go pointer references to addressesToData that were referenced by the encoder
	//  into C ddwaf_objects. libddwaf's API requires to keep this data for the lifetime of the ddwaf_context.
	defer context.cgoRefs.append(encoder.cgoRefs)
```

File: context.go
Comment (Julio-Guerra) [inline context.go:121]: 1. I don't see any value for `closeContext` which spreads to context closure into an handle method
2. The `keepAlive` calls don't make sense to me here, because nothing special happens before.

So here is my suggestion:
```suggestion
	defer context.handle.closeContext(context)
	// Needed to make sure the garbage collector does not throw the values send to the WAF earlier than necessary
	wafLib.wafContextDestroy(context.cContext)
	// Keep the Go pointer references until the end of the context
	keepAlive(context.cgoRefs)
	// The context is no longer used so we can try releasing the Go pointer references asap by nulling them
	context.cgoRefs = cgoRefPool{}
```

If applied, renaming `closeContext` sounds better too.

File: ctypes.go
Comment (Julio-Guerra) [inline ctypes.go:136]:
```suggestion
// cast is used to centralize unsafe use C allocated pointer.
```

File: ctypes.go
Comment (Julio-Guerra) [inline ctypes.go:138]: If you want to give it a try, I think this is the solution to accept pointers and knowing the pointer type:

```go
func cast[T any, PT interface { *T }](ptr uintptr) PT
```

**Review by Julio-Guerra — COMMENTED** [review]
New pass (still not finished 😬)

**Review by Hellzy — APPROVED** [review]
I think we've iterated enough for a first version, this looks okay to me. Great job 💪
Let's address the last few comments and wait for @Julio-Guerra's approval 👍

File: context.go
Comment (eliottness) [reply context.go:121]: I was against the idea because of owernship. Just like why NewContext() is within the handle scope, I think that destroying the context should also be done via the handle. This idea already became a little flaky after you told me to change NewContext() from a method to a function but yeah

File: ctypes.go
Comment (eliottness) [reply ctypes.go:138]: image

File: context.go
Comment (Julio-Guerra) [reply context.go:124]: Part of this proposal still apply even with closeContext:
```suggestion
	defer context.handle.closeContext(context)
	// Keep the Go pointer references until the end of the context
	keepAlive(context.cgoRefs)
	// The context is no longer used so we can try releasing the Go pointer references asap by nulling them
	context.cgoRefs = cgoRefPool{}
```

File: handle.go
Comment (Julio-Guerra) [inline handle.go:71]:
```suggestion
		return nil, errors.New("could not instantiate the WAF")
```

**Review by Julio-Guerra — CHANGES_REQUESTED** [review]

File: handle.go
Comment (eliottness) [reply handle.go:118]: Do you want me to move it ? I would need to rename everything since it does not only decrementReferences since it also closes 👀

Comment (Julio-Guerra) [reply handle.go:118]: I'm good, simple comment to emphasis it.

**Review by Julio-Guerra — APPROVED** [review]

**Pull request merged**
